src/x3d/x3dloadinternalstarling.pas

```diff
@@ -0,0 +1,611 @@
+            TexCoordInterp := TCoordinateInterpolator2DNode.Create(LastAnimationName + '_TexCoord');
+
+            AddFrameCoords(CoordInterp, TexCoordInterp);
+          end else
+          begin
+            { Next frame of animation }
+            Inc(CurrentAnimFrameCount);
+
+            AddFrameCoords(CoordInterp, TexCoordInterp);
+          end;
+        end;
+
+        { Add last animation }
+        if CurrentAnimFrameCount > 0 then
+          AddAnimation(CurrentAnimFrameCount, TimeSensor, CoordInterp, TexCoordInterp);
+
+        Result := FRoot;
+
+      finally
+        FreeAndNil(I);
+      end;
+    except
+      FreeAndNil(FRoot);
+      raise;
+    end;
+  finally
+    FreeAndNil(Doc);
+  end;
+end;
+
+destructor TStarlingTextureAtlasLoader.Destroy;
+begin
+  FreeAndNil(FSubTexture);
+  FreeAndNil(FAnimationList);
+  inherited Destroy;
+end;
+
+{ TStarlingSubTexture }
+
+procedure TStarlingTextureAtlasLoader.TStarlingSubTexture.PrepareTexCordsForX3D(
+  const ImageWidth, ImageHeight: Integer);
+begin
+  { The input data (X1, Y1) are the coordinates in the texture.
+    We need those coordinates to compute X2, Y2 and after that we
+    recalculate X1, X2 for X3D. }
+  X2 := 1 / ImageWidth * (X1 + Width);
+  Y2 := 1 - 1 / ImageHeight * (Y1 + Height);
+
+  X1 := 1 / ImageWidth * X1;
+  Y1 := 1 - 1 / ImageHeight * Y1;
+end;
+
+procedure TStarlingTextureAtlasLoader.TStarlingSubTexture.ParseAnimationName(
+  const SubTextureName: String);
+var
+  UnderscorePos: SizeInt;
+  FrameNumberStr: String;
+  FrameNumber: Integer;
+  AnimationNameLength: Integer;
+begin
+  { Some files has file type Extensions like "walk/0001.png" or walk_1.png }
+  AnimationName := DeleteFileExt(SubTextureName);
+
+  case FAnimationNaming of
+    anStrictUnderscore:
+      begin
+        UnderscorePos := rpos('_', AnimationName);
+
+        { Check characters after underscore is number if not don't change name. }
+        AnimationNameLength := Length(AnimationName);
+        if (UnderscorePos > 0) and (AnimationNameLength > UnderscorePos) then
+        begin
+          FrameNumberStr := Copy(AnimationName, UnderscorePos + 1, AnimationNameLength - UnderscorePos);
+          if TryStrToInt(FrameNumberStr,  FrameNumber) then
+          begin
+            Delete(AnimationName, UnderscorePos, AnimationNameLength - UnderscorePos + 1);
+          end;
+        end;
+      end;
+    anTralingNumber:
+      begin
+        RemoveTrailingChars(AnimationName, ['0'..'9']);
+
+        { Remove trailing underscore if there is more than one char }
+        AnimationNameLength := Length(AnimationName);
+        if (AnimationNameLength > 1) and (AnimationName[AnimationNameLength] = '_') then
+          delete(AnimationName, AnimationNameLength, 1);
+      end;
+  end;
+
+  if AnimationName = '' then
+  begin
+    WritelnWarning('Starling', 'Incorrect animation name (%s), I set to "unknown"',
+    [SubTextureName]);
```

Comment:
Indent the line `[SubTextureName]);` with 2 spaces, i.e.

```pascal
WritelnWarning('Starling', 'Incorrect animation name (%s), I set to "unknown"',
  [SubTextureName]);
```

Another way to write some Format-like lines, that I also use (but both above and below are correct):

```pascal
WritelnWarning('Starling', 'Incorrect animation name (%s), I set to "unknown"', [
  SubTextureName
]);
```

I found it is easier to add / remove / comment-out particular parameters is the last form.